File: Example/JWTDesktop/JWTDesktop/ViewController.m
//
//  ViewController.m
//  JWTDesktop
//
//  Created by Lobanov Dmitry on 23.05.16.
//  Copyright © 2016 JWT. All rights reserved.
//

#import "ViewController.h"
#import <JWT/JWT.h>
#import <JWT/JWTAlgorithmFactory.h>
#import <JWT/JWTAlgorithmNone.h>
#import <Masonry/Masonry.h>
#import "JWTTokenTextTypeDescription.h"
#import "SignatureValidationDescription.h"
#import "JWTDecriptedViewController.h"
#import "ViewController+Model.h"

@interface ViewController() <NSTextViewDelegate, NSTextFieldDelegate, NSTableViewDelegate, NSTableViewDataSource>
@property (weak) IBOutlet NSTextField *algorithmLabel;
@property (weak) IBOutlet NSPopUpButton *algorithmPopUpButton;
@property (weak) IBOutlet NSTextField *secretLabel;
@property (weak) IBOutlet NSTextField *secretTextField;
@property (weak) IBOutlet NSButton *secretIsBase64EncodedCheckButton;

@property (unsafe_unretained) IBOutlet NSTextView *encodedTextView;
@property (unsafe_unretained) IBOutlet NSTextView *decodedTextView;
@property (weak) IBOutlet NSTableView *decodedTableView;
@property (weak) IBOutlet NSView * decriptedView;
@property (strong, nonatomic, readwrite) JWTDecriptedViewController *decriptedViewController;
@property (weak) IBOutlet NSTextField *signatureStatusLabel;

@property (strong, nonatomic, readwrite) ViewController__Model *model;
@end

// it catches all data from view controller

@interface ViewController (JWTTokenDecoderNecessaryDataObject__Protocol) <JWTTokenDecoderNecessaryDataObject__Protocol>
@end

@implementation ViewController (JWTTokenDecoderNecessaryDataObject__Protocol)
- (NSString *)chosenAlgorithmName {
    return [self.algorithmPopUpButton selectedItem].title;
}

- (NSData *)chosenSecretData {
    NSString *secret = [self chosenSecret];
    
    BOOL isBase64Encoded = [self isBase64EncodedSecret];
    NSData *result = nil;
    
    if (isBase64Encoded) {
        result = [[NSData alloc] initWithBase64EncodedString:secret options:0];
        if (!result) {
            self.secretIsBase64EncodedCheckButton.integerValue = 0;
        }
    }
    
    return result;
}

- (NSString *)chosenSecret {
    return self.secretTextField.stringValue;
}

- (BOOL)isBase64EncodedSecret {
    return self.secretIsBase64EncodedCheckButton.integerValue == 1;
}
@end


@implementation ViewController

#pragma mark - Refresh UI
- (void)refreshUI {
    
    NSTextStorage *textStorage = self.encodedTextView.textStorage;
    NSString *string = textStorage.string;
    NSAttributedString *attributedString = [self.model.tokenAppearance encodedAttributedTextForText:string serialization:self.model.tokenSerialization tokenDescription:self.model.tokenDescription];
    NSRange range = NSMakeRange(0, string.length);
    
//    [self.encodedTextView insertText:attributedString replacementRange:range];
    [self.encodedTextView.undoManager beginUndoGrouping];
    [textStorage replaceCharactersInRange:range withAttributedString:attributedString];
    [self.encodedTextView.undoManager endUndoGrouping];
    
    NSError *error = nil;
    NSDictionary *result = [self.model.decoder decodeToken:string skipSignatureVerification:YES error:&error necessaryDataObject:self];
    
    NSString *decodedTokenAsJSON = [self.model.tokenSerialization stringFromDecodedToken:result];
    BOOL signatureVerified = [self.model.decoder decodeToken:string skipSignatureVerification:NO error:&error necessaryDataObject:self]!=nil;
    [self signatureReactOnVerifiedToken:signatureVerified];
    
    // will be udpated.
    JWTCodingResultType *resultType = error ? [[JWTCodingResultType alloc] initWithErrorResult:[[JWTCodingResultTypeError alloc] initWithError:error]] : [[JWTCodingResultType alloc] initWithSuccessResult:[[JWTCodingResultTypeSuccess alloc] initWithHeaders:result[JWTCodingResultHeaders] withPayload:result[JWTCodingResultPayload]]];
    self.decriptedViewController.resultType = resultType;
    // not used.
    [self.decodedTextView replaceCharactersInRange:range withString:decodedTokenAsJSON];
}

#pragma mark - Signature Customization
- (void)signatureReactOnVerifiedToken:(BOOL)verified {
    SignatureValidationType type = verified ? SignatureValidationTypeValid : SignatureValidationTypeInvalid;
    self.model.signatureValidationDescription.signatureValidation = type;
    self.signatureStatusLabel.textColor = self.model.signatureValidationDescription.currentColor;
    self.signatureStatusLabel.stringValue = self.model.signatureValidationDescription.currentTitle;
}

#pragma mark - Setup
- (void)setupModel {
    self.model = [ViewController__Model new];
}

- (void)setupTop {
    // top label.
    self.algorithmLabel.stringValue = @"Algorithm";
    
    // pop up button.
    [self.algorithmPopUpButton removeAllItems];
    [self.algorithmPopUpButton addItemsWithTitles:self.model.availableAlgorithmsNames];
    [self.algorithmPopUpButton setAction:@selector(popUpButtonValueChanged:)];
    [self.algorithmPopUpButton setTarget:self];
    
    // secretLabel
    self.secretLabel.stringValue = @"Secret";
    
    // secretTextField
    self.secretTextField.placeholderString = @"Secret";
    self.secretTextField.delegate = self;
    
    // check button
    self.secretIsBase64EncodedCheckButton.title = @"is Base64Encoded Secret";
    self.secretIsBase64EncodedCheckButton.integerValue = NO;
    [self.secretIsBase64EncodedCheckButton setTarget:self];
    [self.secretIsBase64EncodedCheckButton setAction:@selector(checkBoxState:)];
}

- (void)setupBottom {
    self.signatureStatusLabel.alignment       = NSTextAlignmentCenter;
    self.signatureStatusLabel.textColor       = [NSColor whiteColor];
    self.signatureStatusLabel.drawsBackground = YES;
    
    self.model.signatureValidationDescription.signatureValidation = SignatureValidationTypeUnknown;
    self.signatureStatusLabel.textColor = self.model.signatureValidationDescription.currentColor;
    self.signatureStatusLabel.stringValue = self.model.signatureValidationDescription.currentTitle;
}

- (void)setupEncodingDecodingViews {
    self.encodedTextView.delegate = self;
//    self.decodedTextView.delegate = self;
    self.decodedTableView.delegate = self;
    self.decodedTableView.dataSource = self;
    
    //thanks!
    //http://stackoverflow.com/questions/7545490/how-can-i-have-the-only-column-of-my-nstableview-take-all-the-width-of-the-table
    NSTableView *tableView = self.decodedTableView;
    [tableView  setColumnAutoresizingStyle:NSTableViewUniformColumnAutoresizingStyle];
    [tableView.tableColumns.firstObject setResizingMask:NSTableColumnAutoresizingMask];
    //AND
    [tableView sizeLastColumnToFit];
}

- (void)setupDecorations {
    [self setupTop];
    [self setupBottom];
}

- (void)setupDecriptedViews {
    NSView *view = self.decriptedView;
    self.decriptedViewController = [JWTDecriptedViewController new];
    [view addSubview:self.decriptedViewController.view];
    // maybe add contstraints.
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupModel];
    [self setupDecorations];
    [self setupEncodingDecodingViews];
    [self setupDecriptedViews];
    [self defaultDataSetup];
    [self refreshUI];
    // Do any additional setup after loading the view.
}
- (void)defaultDataSetup {
    ViewController__DataSeed *dataSeed = [ViewController__DataSeed RS256];
    [self defaultDataSetupWithToken:dataSeed.token secret:dataSeed.secret algorithmName:dataSeed.algorithmName];
}

- (void)defaultDataSetupWithToken:(NSString *)token secret:(NSString *)secret algorithmName:(NSString *)algorithmName {
    if (token == nil || secret == nil || algorithmName == nil) {
        NSLog(@"%@ failed! one of them is nil: token:(%@) secret(%@) algorithmName:(%@)algorithm", NSStringFromSelector(_cmd), token, secret, algorithmName);
        return;
    }
    // token
    [self.encodedTextView insertText:token replacementRange:NSMakeRange(0, token.length)];
    
    // secret
    self.secretTextField.stringValue = secret;
    
    // algorithm
    NSInteger index = [self.model.availableAlgorithmsNames indexOfObject:algorithmName];
    [self.algorithmPopUpButton selectItemAtIndex:index];
}

- (void)viewWillAppear {
    [super viewWillAppear];
    NSView *view = self.decriptedView;
    [self.decriptedViewController.view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(view);
    }];
}

#pragma mark - Actions
- (void)popUpButtonValueChanged:(id)sender {
    [self refreshUI];
}

-(IBAction)checkBoxState:(id)sender {
    // Under construction
    [self refreshUI];
}


#pragma marrk - Delegates / <NSTextFieldDelegate>

- (void)controlTextDidChange:(NSNotification *)obj {
    if ([obj.name isEqualToString:NSControlTextDidChangeNotification]) {
        NSTextField *textField = (NSTextField *)obj.object;
        if (textField == self.secretTextField) {
            // refresh UI
            [self refreshUI];
        }
    }
}

#pragma mark - EncodedTextView / <NSTextViewDelegate>

- (void)textDidChange:(NSNotification *)notification {
    [self refreshUI];
}

#pragma mark - DecodedTableView / <NSTableViewDataSource>

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return 4;
}

#pragma mark - DecodedTableView / <NSTableViewDelegate>
- (BOOL)tableView:(NSTableView *)tableView isGroupRow:(NSInteger)row {
    return row % 2 == 0;
}
- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    // choose by row is section or not
    if (row % 2) {
        // section
        NSView *cell = [tableView makeViewWithIdentifier:@"Cell" owner:self];
        ((NSTableCellView *)cell).textField.stringValue = @"AH";
        return cell;
    }
    else {
        NSView *cell = [tableView makeViewWithIdentifier:@"Cell" owner:self];
        ((NSTableCellView *)cell).textField.stringValue = @"OH";
        //    return nil;
        return cell;
    }
}

- (CGFloat)tableView:(NSTableView *)tableView heightOfRow:(NSInteger)row {
    // calculate height of row.
//    NSView * view = [tableView viewAtColumn:0 row:row makeIfNecessary:NO];
    return 40;
}

@end

File: Example/JWTDesktopSwift/JWTDesktopSwift/ViewController.swift
//
//  ViewController.swift
//  JWTDesktopSwift
//
//  Created by Lobanov Dmitry on 01.10.16.
//  Copyright © 2016 JWTIO. All rights reserved.
//

import Cocoa
import JWT
import JWTDesktopSwiftToolkit


// MARK: - Supply JWT Methods
extension ViewController {
    func tokenDataTransferObject() -> TokenDecoder.DataTransferObject {
        let algorithmName = self.algorithmPopUpButton.selectedItem?.title ?? ""
        let secret = self.secretTextField.stringValue
        let secretData: Data? = self.getSecretData
        let isBase64EncodedSecret = self.secretIsBase64EncodedCheckButton.integerValue == 1
        let shouldSkipSignatureVerification = self.signatureVerificationCheckButton.integerValue == 1
        return .init(algorithmName: algorithmName, secret: secret, secretData: secretData, isBase64EncodedSecret: isBase64EncodedSecret, shouldSkipSignatureVerification: shouldSkipSignatureVerification)
    }
    
    func tokenDataTransferObjectShouldCheckSignature() -> TokenDecoder.DataTransferObject {
        var result = self.tokenDataTransferObject()
        result.shouldSkipSignatureVerification = false
        return result
    }
    
    var getSecretData: Data? {
        let secret = self.secretTextField.stringValue
        
        let isBase64Encoded = self.secretIsBase64EncodedCheckButton.integerValue == 1
        guard let result = Data(base64Encoded: secret), isBase64Encoded else {
            self.secretIsBase64EncodedCheckButton.integerValue = 0
            return nil
        }
        
        return result
    }
}

// Refresh UI
extension ViewController {
    // MARK: - Encoded Text View
    func encodedTextAttributes(_ enumerate: (NSRange, [NSAttributedString.Key : Any]) -> ()) {
        let textStorage = self.encodedTextView.textStorage!
        let string = textStorage.string
        let range = NSMakeRange(0, string.count)
        if let attributedString = self.model.appearance.encodedAttributedString(text: string) {
            attributedString.enumerateAttributes(in: range, options: []) { (attributes, range, bool) in
                enumerate(range, attributes)
            }
        }
    }
    
    // MARK: - Refresh UI
    func refreshUI() {
        
        let textStorage = self.encodedTextView.textStorage!;
        let string = textStorage.string
        self.encodedTextAttributes { (range, attributes) in
            textStorage.setAttributes(attributes, range: range)
        }
        
        // We should add an option to skip verification in decoding section.
        // invalid signature doesn't mean that you can't decode JWT.
        
        if let jwtVerified = self.model.decoder.decode(token: string, object: self.tokenDataTransferObjectShouldCheckSignature()) {
            let notVerified = jwtVerified.successResult?.headerAndPayloadDictionary?.isEmpty == true
            self.signatureReactOnVerifiedToken(verified: !notVerified)
        }
        else {
            self.signatureReactOnVerifiedToken(verified: false)
        }
        
        let result = self.model.decoder.decode(token: string, object: self.tokenDataTransferObject())
        
        self.decriptedViewController.resultType = result
    }
    
    func refreshSignature() {
        self.signatureStatusLabel.backgroundColor = self.model.signatureValidation.color
        self.signatureStatusLabel.stringValue = self.model.signatureValidation.title
    }
}

// MARK: - Actions
extension ViewController {
    @objc func popUpButtonValueChanged(sender : AnyClass) {
        self.refreshUI()
    }
    
    @objc func checkBoxState(sender : AnyClass) {
        self.refreshUI()
    }
    func signatureReactOnVerifiedToken(verified: Bool) {
        self.model.signatureValidation = verified ? .valid : .invalid
        self.refreshSignature()
    }
}

extension ViewController: NSTextFieldDelegate {
    func controlTextDidChange(_ obj: Notification) {
        if (obj.name == NSControl.textDidChangeNotification) {
            let textField = obj.object as! NSTextField
            if textField == self.secretTextField {
                self.refreshUI()
            }
        }
    }
}

extension ViewController: NSTextViewDelegate {
    func textDidChange(_ notification: Notification) {
        self.refreshUI()
    }
//    func textViewDidChangeTypingAttributes(_ notification: Notification) {
//        self.updateEncodedTextAttributes()
//    }
//    func textView(_ textView: NSTextView, shouldChangeTypingAttributes oldTypingAttributes: [String : Any] = [:], toAttributes newTypingAttributes: [NSAttributedString.Key : Any] = [:]) -> [NSAttributedString.Key : Any] {
//        return newTypingAttributes
//    }
    
//    func textView(_ textView: NSTextView, shouldChangeTextIn affectedCharRange: NSRange, replacementString: String?) -> Bool {
//        if (textView == self.encodedTextView) {
////            if let textStore = textView.textStorage {
////                textView.undoManager?.beginUndoGrouping()
////                textStore.replaceCharacters(in: affectedCharRange, with: replacementString!)
////                self.encodedTextAttributes { (range, attributes) in
////                    textStore.setAttributes(attributes, range: range)
////                }
////                textView.undoManager?.endUndoGrouping()
////            }
////            self.refreshUI()
//            return true
//        }
//        return false
//    }
}

// MARK: - EncodingTextViewDelegate
//extension ViewController : NSTextViewDelegate {
//    func textView(_ textView: NSTextView, shouldChangeTextIn affectedCharRange: NSRange, replacementString: String?) -> Bool {
//        if (textView == self.encodedTextView) {
//            if let textStore = textView.textStorage {
//                textStore.replaceCharacters(in: affectedCharRange, with: replacementString!)
//            }
//            self.refreshUI()
//            return false
//        }
//        return false
//    }
//}

class ViewController: NSViewController {
    override init(nibName nibNameOrNil: NSNib.Name?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    // MARK: - Properties - Outlets
    @IBOutlet weak var algorithmLabel : NSTextField!
    @IBOutlet weak var algorithmPopUpButton : NSPopUpButton!
    
    @IBOutlet weak var secretLabel : NSTextField!
    @IBOutlet weak var secretTextField : NSTextField!
    @IBOutlet weak var secretIsBase64EncodedCheckButton : NSButton!
    
    @IBOutlet weak var signatureLabel : NSTextField!
    @IBOutlet weak var signatureVerificationCheckButton : NSButton!
    
    
    @IBOutlet weak var encodedTextView : NSTextView!
    @IBOutlet weak var decriptedView : NSView!
    var decriptedViewController : DecriptedViewController!
    
    @IBOutlet weak var signatureStatusLabel : NSTextField!
    
    // MARK: - Model
    var model: Model!
    
    // MARK: - Setup
    func setupModel() {
        self.model = Model()
    }
    
    func setupTop() {
        // top label.
        self.algorithmLabel.stringValue = "Algorithm";
        // pop up button.
        
        self.algorithmPopUpButton.removeAllItems()
        self.algorithmPopUpButton.addItems(withTitles: self.model.availableAlgorithmsNames)
        self.algorithmPopUpButton.target = self
        self.algorithmPopUpButton.action = #selector(ViewController.popUpButtonValueChanged(sender:))
        
        // secretLabel
        self.secretLabel.stringValue = "Secret"
        
        // secretTextField
        self.secretTextField.placeholderString = "Secret"
        self.secretTextField.delegate = self

        // check button
        self.secretIsBase64EncodedCheckButton.title = "is Base64Encoded Secret"
        self.secretIsBase64EncodedCheckButton.integerValue = 0
        self.secretIsBase64EncodedCheckButton.target = self
        self.secretIsBase64EncodedCheckButton.action = #selector(ViewController.checkBoxState(sender:))
        
        // signatureLabel
        self.signatureLabel.stringValue = "Signature"
        
        // signatureVerificationCheckButton
        self.signatureVerificationCheckButton.title = "Skip signature verification"
        self.signatureVerificationCheckButton.integerValue = 0
        self.signatureVerificationCheckButton.target = self
        self.signatureVerificationCheckButton.action = #selector(ViewController.checkBoxState(sender:))
    }
    
    func setupBottom() {
        self.signatureStatusLabel.alignment       = .center
        self.signatureStatusLabel.textColor       = NSColor.white
        self.signatureStatusLabel.drawsBackground = true
        self.refreshSignature()
    }

    func setupEncodingDecodingViews() {
        self.encodedTextView.delegate = self;
    }
    
    func setupDecorations() {
        self.setupTop()
        self.setupBottom()
    }
    
    func setupDecriptedViews() {
        let view = self.decriptedView
        self.decriptedViewController = DecriptedViewController()
        view?.addSubview(self.decriptedViewController.view)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setupModel()
        
        self.setupDecorations()
        self.setupEncodingDecodingViews()
        self.setupDecriptedViews()
        
        self.defaultDataSetup()
    }
    
    func defaultDataSetup(algorithmName: String, secret: String, token: String) {
        // algorithm HS256
        if let index = self.model.availableAlgorithmsNames.firstIndex(where: {
            $0 == algorithmName
        }) {
            self.algorithmPopUpButton.selectItem(at: index)
        }

        // secret
        self.secretTextField.stringValue = secret
        
        // token
        var range = NSRange()
        range.location = 0
        range.length = token.count
        self.encodedTextView.insertText(token, replacementRange: range)

    }
    
    func defaultDataSetup() {
        let seed = DataSeedType.rs256.dataSeed//rs256.dataSeed
        self.defaultDataSetup(algorithmName: seed.algorithmName, secret: seed.secret, token: seed.token)
    }
    
    override func viewWillAppear() {
        super.viewWillAppear()
        guard let view = self.decriptedView else { return }
        
        let subview = self.decriptedViewController.view

        view.translatesAutoresizingMaskIntoConstraints = false
        subview.translatesAutoresizingMaskIntoConstraints = false
        
        let constraints = [
            subview.leftAnchor.constraint(equalTo: view.leftAnchor),
            subview.rightAnchor.constraint(equalTo: view.rightAnchor),
            subview.topAnchor.constraint(equalTo: view.topAnchor),
            subview.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ]
        
        NSLayoutConstraint.activate(constraints)
    }
}

